cupyx/scipy/sparse/linalg/solve.py

```diff
@@ -65,3 +66,237 @@ def lsqr(A, b):
+    descr_U = cupy.cusparse.MatDescriptor.create()
+    descr_M.set_mat_index_base(cusparse.CUSPARSE_INDEX_BASE_ONE)
+    descr_M.set_mat_type(cusparse.CUSPARSE_MATRIX_TYPE_GENERAL)
+    descr_L.set_mat_index_base(cusparse.CUSPARSE_INDEX_BASE_ONE)
+    descr_L.set_mat_type(cusparse.CUSPARSE_MATRIX_TYPE_GENERAL)
+    descr_L.set_mat_fill_mode(cusparse.CUSPARSE_FILL_MODE_LOWER)
+    descr_L.set_mat_diag_type(cusparse.CUSPARSE_DIAG_TYPE_UNIT)
+    descr_U.set_mat_index_base(cusparse.CUSPARSE_INDEX_BASE_ONE)
+    descr_U.set_mat_type(cusparse.CUSPARSE_MATRIX_TYPE_GENERAL)
+    descr_U.set_mat_fill_mode(cusparse.CUSPARSE_FILL_MODE_UPPER)
+    descr_U.set_mat_diag_type(cusparse.CUSPARSE_DIAG_TYPE_NON_UNIT)
+
+    # get required buffer size and allocate corresponding memory
+    buff_size_M = cupy.cusparse._call_cusparse(
+        'csrilu02_bufferSize', dtype, handle, *A_tuple(descr_M), info_M)
+    buff_size_L = cupy.cusparse._call_cusparse(
+        'csrsv2_bufferSize', dtype, handle, trans_L, *A_tuple(descr_L), info_L)
+    buff_size_U = cupy.cusparse._call_cusparse(
+        'csrsv2_bufferSize', dtype, handle, trans_U, *A_tuple(descr_U), info_U)
+    buff_size = numpy.max((buff_size_M, buff_size_L, buff_size_U))
+    buff = cupy.empty(int(buff_size), dtype=cupy.int8)
+    assert buff.data.ptr % 128 == 0
+
+    # Analysis (setup) of M = LU, the Incomplete LU factorization of A
+    cupy.cusparse._call_cusparse('csrilu02_analysis', dtype, handle,
+                                 *A_tuple(descr_M), info_M, policy_M,
+                                 buff.data.ptr)
+    cupy.cusparse._call_cusparse('csrsv2_analysis', dtype, handle, trans_L,
+                                 *A_tuple(descr_L), info_L, policy_L,
+                                 buff.data.ptr)
+    cupy.cusparse._call_cusparse('csrsv2_analysis', dtype, handle, trans_U,
+                                 *A_tuple(descr_U), info_U, policy_U,
+                                 buff.data.ptr)
+
+    # Perform M = L * U incomplete decomposition
+    cupy.cusparse._call_cusparse('csrilu02', dtype, handle,
+                                 *A_tuple(descr_M), info_M, policy_M,
+                                 buff.data.ptr)
+
+    y = cupy.empty(n, dtype=dtype)
+    out = cupy.empty(n, dtype=dtype)
+
+    def M(x):
+        cupy.cusparse._call_cusparse('csrsv2_solve', dtype, handle, trans_L,
+                                     *A_tuple_a(descr_L), info_L, x.data.ptr,
+                                     y.data.ptr, policy_L, buff.data.ptr)
+        cupy.cusparse._call_cusparse('csrsv2_solve', dtype, handle, trans_U,
+                                     *A_tuple_a(descr_U), info_U, y.data.ptr,
+                                     out.data.ptr, policy_U, buff.data.ptr)
+        return out
+
+    return M, (info_M, info_U, info_L)
+
+
+def _destroy_ilu(info_list):
+    if info_list is not None:
+        for info in info_list:
+            cusparse.destroyCsrilu02Info(info)
```

Comment:
The type of `info` might be `Csrsv2Info`. When it is `Csrsv2Info`, `cusparse.destroyCsrsv2Info()` needs to be used as destructor.